### PR TITLE
Rework the "Using Plutus Tx" section of the user guide

### PR DESCRIPTION
**Pull request by zliu41**
The section now looks like this:

Screenshot from 2024-09-24 11-24-38

- "Optimizing scripts with `asData`": unmodified
- "GHC Extensions, Flags and Pragmas": moved a small portion to "Differences From Haskell"
- All other pages are written from scratch.

### CODE REVIEW

**Comment by effectfully** [comment]
(sorry for hijacking)

File: doc/docusaurus/docs/using-plutus-tx/compiling-plutus-tx.md [hunk 1]
```diff
@@ -1,95 +1,67 @@
...
-All functions and methods should have the `INLINEABLE` pragma, so that their unfoldings are made available to the Plutus Tx compiler.
+import Data.Proxy
+import PlutusTx.Code (CompiledCode)
+import PlutusTx.Plugin (plc)
 
-The `-fexpose-all-unfoldings` flag also makes GHC expose all unfoldings, but unfoldings exposed this way can be more optimized than unfoldings exposed via `INLINEABLE`.
-In general, we do not want GHC to perform optimizations, since GHC optimizes a program based on the assumption that it has non-strict semantics and is evaluated lazily (call by need), which is not true for Plutus Tx programs.
-Therefore, `INLINEABLE` is preferred over `-fexpose-all-unfoldings`, even though the latter is simpler.
+myPlutusTxCodeCompiled :: CompiledCode (Integer -> Integer)
+myPlutusTxCodeCompiled = plc (Proxy @"location info") myPlutusTxCode
```
Comment (bezirg) [inline doc/docusaurus/docs/using-plutus-tx/compiling-plutus-tx.md:64]: maybe mention that "location info" is for debugging purposes and not very important

File: doc/docusaurus/docs/using-plutus-tx/extensions-flags-pragmas.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+>
+> This section only covers GHC flags, not Plutus Tx compiler flags.
+> A number of options can be passed to the Plutus Tx compiler.
+> See [Plutus Tx Compiler Options](../delve-deeper/plutus-tx-compiler-options.md) for details.
+
+### Pragmas
+
+All functions and methods should have the `INLINEABLE` pragma, so that their unfoldings are made available to the Plutus Tx compiler.
+
+The `-fexpose-all-unfoldings` flag also makes GHC expose all unfoldings, but unfoldings exposed this way can be more optimized than unfoldings exposed via `INLINEABLE`.
```
Comment (bezirg) [inline doc/docusaurus/docs/using-plutus-tx/extensions-flags-pragmas.md:73]: I would start with "Alternatively, ..."

File: doc/docusaurus/docs/using-plutus-tx/special-functions-and-types.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+It is useful to be aware of these, as these functions and types work differently than regular user-defined functions and types, and cannot be defined in user space.
+
+## Builtin Functions and Types
+
+There are a number of builtin functions and builtin types in UPLC.
+The list of builtin functions and types can be found in the [Plutus Core Specification](https://plutus.cardano.intersectmbo.org/resources/plutus-core-spec.pdf).
+
+In [`PlutusTx.Builtins.Internal`](https://plutus.cardano.intersectmbo.org/haddock/latest/plutus-tx/PlutusTx-Builtins-Internal.html), functions marked `OPAQUE` are directly converted to the corresponding builtin function.
+For instance, `PlutusTx.Builtins.Internal.addInteger` is converted to UPLC term `(builtin addInteger)`.
+Using the `OPAQUE` pramga prevents GHC from inlining these functions, allowing the Plutus Tx compiler to determine where exactly they are invoked, and compile them appropriately.
```
Comment (bezirg) [inline doc/docusaurus/docs/using-plutus-tx/special-functions-and-types.md:21]: typo "pragma"

File: doc/docusaurus/docs/using-plutus-tx/special-functions-and-types.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+## `PlutusTx.Bool.&&` and `PlutusTx.Bool.||`
+
+In [`PlutusTx.Bool`](https://plutus.cardano.intersectmbo.org/haddock/latest/plutus-tx/PlutusTx-Bool.html), the `&&` and `||` operators are handled specially in the plugin to ensure they can short-circuit.
+In most strict languages, these operators are special and cannot be defined by users, which is also the case for Plutus Tx.
+
+Note that in regular Haskell, `&&` and `||` are _not_ special, as Haskell supports non-strict function applications (and it is the default behavior).
+
+## `IsString` Instances
+
+[`IsString`](https://hackage.haskell.org/package/base/docs/Data-String.html#t:IsString) is a type class from `base`, and can be used along with the `OverloadedString` language extension to convert string literals to types other than `String`.
```
Comment (bezirg) [inline doc/docusaurus/docs/using-plutus-tx/special-functions-and-types.md:42]: typo "OverloadedStrings"

**Review by bezirg — APPROVED** [review]

**Pull request merged**